src/providers/sh/util/build-logger.js

```diff
@@ -25,6 +25,13 @@ module.exports = class Logger extends EventEmitter {
     this.socket.on('logs', this.onLog.bind(this))
     this.socket.on('backend', this.onComplete.bind(this))
 
+    this.socket.on('disconnect', () => {
+      // ensure close event fires even if already built
+      if (!this.building) {
+        this.emit('close')
+      }
+    })
+
```

Comment:
The logger depends on the `close` event firing in order to exit, which wasn't happening before in the case of the deployment already being built (i.e. `state` is never triggered). This makes sure it is called in this case.